packages/amd-loader/src/amd-loader.ts

```diff
@@ -0,0 +1,242 @@
+
+class Module {
+  private url: normalizedUrl;
+  private urlBase: string;
+  private exports: {[id: string]: {}} = {};
+
+  /** All dependencies are resolved and the factory has run. */
+  private resolved = false;
+
+  /** Callbacks from dependents waiting for this module to resolve. */
+  private notify: Array<() => void> = [];
+
+  public needsLoad = true;
+
+  constructor(url: string) {
+    this.url = normalizeUrl(url);
+    this.urlBase = getBasePath(this.url);
+  }
+
+  /**
+   * Initialize this module with its dependencies and factory function. Note
+   * that Module objects are created and registered before they are loaded,
+   * which is why this is not simply part of construction.
+   */
+  define(deps: string[], factory?: (...args: {}[]) => void) {
+    const mod = this;
+    this.require(deps, function(...args: {}[]) {
+      if (factory !== undefined) {
+        factory.apply(null, args);
+      }
+      mod.resolved = true;
+      for (const callback of mod.notify) {
+        callback();
+      }
+    });
+  }
+
+  /**
+   * Execute the given callback after all dependencies are resolved.
+   */
+  require(deps: string[], callback?: (...args: {}[]) => void) {
+    const args: {}[] = [];
+    let numUnresolvedDeps = 0;
+
+    function onDepResolved() {
+      numUnresolvedDeps--;
+      checkDeps();
+    }
+
+    function checkDeps() {
+      if (numUnresolvedDeps === 0) {
+        if (callback !== undefined) {
+          callback.apply(null, args);
+        }
+      }
+    }
+
+    for (let i = 0; i < deps.length; i++) {
+      const depSpec = deps[i];
+
+      if (depSpec === 'exports') {
+        args[i] = this.exports;
+
+      } else if (depSpec === 'require') {
+        args[i] = this.require.bind(this);
+
+      } else if (depSpec === 'meta') {
+        // TODO(aomarks) Possibly replace with
+        args[i] = {url: this.url};
+
+      } else {
+        const depMod = getModule(this.resolveUrl(depSpec));
+        args[i] = depMod.exports;
+
+        if (depMod.resolved === false) {
+          numUnresolvedDeps++;
+          depMod.notify.push(onDepResolved);
+          depMod.loadIfNeeded();
+        }
+      }
+    }
+
+    checkDeps();
+  }
+
+  /**
+   * Resolve a URL relative to the URL of this module.
+   */
+  private resolveUrl(url: string) {
+    if (url.indexOf('://') !== -1) {
+      // Already a fully qualified URL.
+      return url;
+    }
+    return normalizeUrl(this.urlBase + url);
+  }
+
+  /**
+   * Load this module by creating a <script> tag in the document <head>, unless
+   * we have already started (or didn't need to, as in the case of top-level
+   * scripts).
+   */
+  private loadIfNeeded() {
+    if (this.needsLoad === false) {
+      return;
+    }
+    this.needsLoad = true;
+
+    const script = document.createElement('script');
+    script.src = this.url;
+
+    const mod = this;
+    script.onload = function() {
+      if (pendingDefine !== undefined) {
+        pendingDefine(mod);
```

Comment:
Discussed offline. The timing doesn't work out -- the module object has already been created by the parent module at this time (whichever one first declared a dependency on this module), and here we are actually waiting to find out our URL so that we can associate that existing module object with the dependencies and factory that we now have.